e2e/pos_test.go

```diff
@@ -0,0 +1,617 @@
+
+	// expBalance = previousAccountBalance + stakeRefund - 1 ETH - block fees
+	expBalance := big.NewInt(0).Sub(big.NewInt(0).Add(defaultBalance, bigDefaultStakedBalance), oneEth)
+	expBalance = big.NewInt(0).Sub(expBalance, paidFee)
+
+	assert.Equalf(t,
+		expBalance.String(),
+		actualAccountBalance.String(),
+		"Account balance mismatch after stake / unstake exploit",
+	)
+}
+
+// Test scenario:
+// User has 0 ETH staked and a balance of 100 ETH
+// Stake 2 ETH -> Unstake
+// Expected result for tests: Staked: 0 ETH; Balance: ~100 ETH
+func TestPoS_StakeUnstakeWithinSameBlock(t *testing.T) {
+	// Predefined values
+	var blockGasLimit uint64 = 5000000000
+	stakingContractAddr := staking.AddrStakingContract
+	defaultBalance := framework.EthToWei(100)
+	bigGasPrice := big.NewInt(framework.DefaultGasPrice)
+
+	senderKey, senderAddr := tests.GenerateKeyAndAddr(t)
+	numDummyStakers := 10
+
+	devInterval := 5 // s
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(devInterval)
+		config.Premine(senderAddr, defaultBalance)
+		config.SetBlockLimit(blockGasLimit)
+		config.SetDevStakingAddresses(generateStakingAddresses(numDummyStakers))
+		config.SetIBFTPoS(true)
+	})
+	srv := srvs[0]
+	client := srv.JSONRPC()
+
+	initialStakingSCBalance, fetchError := framework.GetStakedAmount(senderAddr, client)
+	if fetchError != nil {
+		t.Fatalf("Unable to fetch staking SC balance, %v", fetchError)
+	}
+
+	// Required default values
+	signer := crypto.NewEIP155Signer(100)
+	currentNonce := 0
+
+	// TxPool client
+	txpoolClient := srv.TxnPoolOperator()
+
+	generateTx := func(value *big.Int, methodName string) *types.Transaction {
+		signedTx, signErr := signer.SignTx(&types.Transaction{
+			Nonce:    uint64(currentNonce),
+			From:     types.ZeroAddress,
+			To:       &stakingContractAddr,
+			GasPrice: bigGasPrice,
+			Gas:      framework.DefaultGasLimit,
+			Value:    value,
+			V:        big.NewInt(1), // it is necessary to encode in rlp
+			Input:    framework.MethodSig(methodName),
+		}, senderKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		currentNonce++
+
+		return signedTx
+	}
+
+	oneEth := framework.EthToWei(1)
+	zeroEth := framework.EthToWei(0)
+	for i := 0; i < 2; i++ {
+		var msg *txpoolOp.AddTxnReq
+		if i%2 == 0 {
+			stakeTxn := generateTx(oneEth, "stake")
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: stakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		} else {
+			unstakeTxn := generateTx(zeroEth, "unstake")
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: unstakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		}
+
+		_, addErr := txpoolClient.AddTxn(context.Background(), msg)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
```

Comment:
why the for loop for these 2 transactions? Is it because of the c/p from the previous one? 🙂

Reply:
![](https://media.giphy.com/media/y6Inkaz7omxAk/giphy.gif)

Fixed 👍 

8831ce9